File: Pharmaceutical_Company_Example/Behavioral/R/Mediator.R
# Mediator Interface
Mediator <- setRefClass(
  "Mediator",
  methods = list(
    notify = function(sender, event) {
      stop("This method should be overridden by concrete mediators!")
    }
  )
)

# Concrete Mediator
RegulatoryMediator <- setRefClass(
  "RegulatoryMediator",
  contains = "Mediator",  # <-- Explicit inheritance from Mediator
  fields = list(clinical_trials = "ANY", quality_assurance = "ANY", regulatory_affairs = "ANY", legal = "ANY"),
  methods = list(
    notify = function(sender, event) {
      if (event == "Clinical Trial Completed") {
        message("Mediator: Clinical Trial completed. Notifying Quality Assurance.")
        quality_assurance$check_quality()
      } else if (event == "Quality Check Completed") {
        message("Mediator: Quality Check completed. Notifying Regulatory Affairs.")
        regulatory_affairs$prepare_documents()
      } else if (event == "Documents Prepared") {
        message("Mediator: Documents prepared. Notifying Legal Department.")
        legal$review_documents()
      } else if (event == "Legal Review Completed") {
        message("Mediator: Legal Review completed. Drug is ready for final approval.")
      }
    }
  )
)

# Department Interfaces
Department <- setRefClass(
  "Department",
  fields = list(mediator = "Mediator"),
  methods = list(
    complete_task = function() {
      stop("This method should be overridden by concrete departments!")
    }
  )
)

# Concrete Department: Clinical Trials
ClinicalTrials <- setRefClass(
  "ClinicalTrials",
  contains = "Department",
  methods = list(
    run_trials = function() {
      message("Clinical Trials: Running clinical trials...")
      mediator$notify(.self, "Clinical Trial Completed")
    }
  )
)

# Concrete Department: Quality Assurance
QualityAssurance <- setRefClass(
  "QualityAssurance",
  contains = "Department",
  methods = list(
    check_quality = function() {
      message("Quality Assurance: Checking product quality...")
      mediator$notify(.self, "Quality Check Completed")
    }
  )
)

# Concrete Department: Regulatory Affairs
RegulatoryAffairs <- setRefClass(
  "RegulatoryAffairs",
  contains = "Department",
  methods = list(
    prepare_documents = function() {
      message("Regulatory Affairs: Preparing regulatory documents...")
      mediator$notify(.self, "Documents Prepared")
    }
  )
)

# Concrete Department: Legal Department
LegalDepartment <- setRefClass(
  "LegalDepartment",
  contains = "Department",
  methods = list(
    review_documents = function() {
      message("Legal Department: Reviewing regulatory documents...")
      mediator$notify(.self, "Legal Review Completed")
    }
  )
)

# Create the Mediator
mediator <- RegulatoryMediator$new()

# Create Departments
clinical_trials <- ClinicalTrials$new(mediator = mediator)
quality_assurance <- QualityAssurance$new(mediator = mediator)
regulatory_affairs <- RegulatoryAffairs$new(mediator = mediator)
legal <- LegalDepartment$new(mediator = mediator)

# Set Departments in the Mediator
mediator$clinical_trials <- clinical_trials
mediator$quality_assurance <- quality_assurance
mediator$regulatory_affairs <- regulatory_affairs
mediator$legal <- legal

# Run process
clinical_trials$run_trials()

File: Pharmaceutical_Company_Example/Behavioral/JavaScript/Mediator.js
// Mediator Interface
class Mediator {
  notify(sender, event) {
    throw new Error("This method should be overridden by concrete mediators!");
  }
}

// Concrete Mediator
class RegulatoryMediator extends Mediator {
  constructor() {
    super();
    this.clinicalTrials = null;
    this.qualityAssurance = null;
    this.regulatoryAffairs = null;
    this.legal = null;
  }

  notify(sender, event) {
    if (event === "Clinical Trial Completed") {
      console.log("Mediator: Clinical Trial completed. Notifying Quality Assurance.");
      if (this.qualityAssurance) {
        this.qualityAssurance.checkQuality();
      }
    } else if (event === "Quality Check Completed") {
      console.log("Mediator: Quality Check completed. Notifying Regulatory Affairs.");
      if (this.regulatoryAffairs) {
        this.regulatoryAffairs.prepareDocuments();
      }
    } else if (event === "Documents Prepared") {
      console.log("Mediator: Documents prepared. Notifying Legal Department.");
      if (this.legal) {
        this.legal.reviewDocuments();
      }
    } else if (event === "Legal Review Completed") {
      console.log("Mediator: Legal Review completed. Drug is ready for final approval.");
    }
  }
}

// Department Interfaces
class Department {
  constructor(mediator) {
    this.mediator = mediator;
  }

  completeTask() {
    throw new Error("This method should be overridden by concrete departments!");
  }
}

// Concrete Department: Clinical Trials
class ClinicalTrials extends Department {
  runTrials() {
    console.log("Clinical Trials: Running clinical trials...");
    this.mediator.notify("ClinicalTrials", "Clinical Trial Completed");
  }
}

// Concrete Department: Quality Assurance
class QualityAssurance extends Department {
  checkQuality() {
    console.log("Quality Assurance: Checking product quality...");
    this.mediator.notify("QualityAssurance", "Quality Check Completed");
  }
}

// Concrete Department: Regulatory Affairs
class RegulatoryAffairs extends Department {
  prepareDocuments() {
    console.log("Regulatory Affairs: Preparing regulatory documents...");
    this.mediator.notify("RegulatoryAffairs", "Documents Prepared");
  }
}

// Concrete Department: Legal Department
class LegalDepartment extends Department {
  reviewDocuments() {
    console.log("Legal Department: Reviewing regulatory documents...");
    this.mediator.notify("LegalDepartment", "Legal Review Completed");
  }
}

// Create the Mediator
const mediator = new RegulatoryMediator();

// Create Departments
const clinicalTrials = new ClinicalTrials(mediator);
const qualityAssurance = new QualityAssurance(mediator);
const regulatoryAffairs = new RegulatoryAffairs(mediator);
const legal = new LegalDepartment(mediator);

// Set Departments in the Mediator
mediator.clinicalTrials = clinicalTrials;
mediator.qualityAssurance = qualityAssurance;
mediator.regulatoryAffairs = regulatoryAffairs;
mediator.legal = legal;

// Run process
clinicalTrials.runTrials();
